python/nav/web/auth.py

```diff
@@ -260,42 +260,54 @@ def login_remote_user(request):
     return None
 
 
+def ensure_account(request):
+    """Guarantee that request.account is set"""
+    session = request.session
+    account = getattr(request, 'account', None)
+
+    # Should only be the case before first page request
+    if ACCOUNT_ID_VAR not in session:
+        session[ACCOUNT_ID_VAR] = Account.DEFAULT_ACCOUNT
+
+    if account is None or account.id != session[ACCOUNT_ID_VAR]:
+        print('here 1', account, session)
+        account = Account.objects.get(id=session[ACCOUNT_ID_VAR])
+        print('here 2', account, session)
+
+    # Called if session[ACCOUNT_ID_VAR] != Account.DEFAULT_ACCOUNT,
+    # since Account.DEFAULT_ACCOUNT should never be locked
+    if account.locked:
+        print('here 3')
+        session[ACCOUNT_ID_VAR] = Account.DEFAULT_ACCOUNT
+        account = Account.objects.get(id=session[ACCOUNT_ID_VAR])
+
+    request.account = account
+
+
 # Middleware
 
 
 class AuthenticationMiddleware(MiddlewareMixin):
     def process_request(self, request):
-        session = request.session
-        account = getattr(request, 'account', None)
-
-        if ACCOUNT_ID_VAR not in session:  # Not logged in
-            # Fallback: Set account id to anonymous user
-            session[ACCOUNT_ID_VAR] = Account.DEFAULT_ACCOUNT
-            # Try remote user
-            login_remote_user(request)
-            account = getattr(request, 'account', None)
-
-        if not account or account.id != session[ACCOUNT_ID_VAR]:
-            # Reget account if:
-            # * account not set
-            # * account.id different from session[ACCOUNT_ID_VAR]
-            account = Account.objects.get(id=session[ACCOUNT_ID_VAR])
+        # Now we have some valid account
+        ensure_account(request)
+        account = request.account
+        sudo_operator = get_sudoer(request)
 
         remote_username = get_remote_username(request)
-        if remote_username and remote_username != account.login:
-            # REMOTE_USER has changed behind your back
-            # Make sure request is correct, for logout's sake
-            request.account = account
-            # Log out current user, log in new user
-            logout(request)
-            login_remote_user(request)
-            account = getattr(request, 'account', None)
-
-        # Now we have an account
-        request.account = account
-
-        if SUDOER_ID_VAR in session:
-            account.sudo_operator = get_sudoer(request)
+        # REMOTE_USER has changed behind our backs
+        if remote_username:
+            if ((sudo_operator and remote_username != sudo_operator.login)
```

Comment:
Go Black! Check out `black-macchiato`. It provides black-formatting of partial code snippets.